lib/Transforms/CombStrengthReduction.cpp

```diff
@@ -0,0 +1,251 @@
+//===- CombStrengthReduction.cpp - A Combinational Strength Reduction pass ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a simplified canonicalizer pass that doesn't do CFG
+// optimizations and other things that aren't helpful for many hardware IRs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Transforms/Passes.h"
+
+#include "circt/Support/LLVM.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace circt;
+using namespace comb;
+
+#define GEN_PASS_CLASSES
+#include "circt/Transforms/Passes.h.inc"
+
+namespace {
```

Comment:
Please only put types in an anon namespace, static is good enough for functions and makes "local reasoning" about the code easier.

Reply:
Done